src/query/plan/variable_start_planner.hpp

```diff
@@ -313,33 +313,77 @@ class VariableStartPlanner {
 
   // Generates different, equivalent query parts by taking different graph
   // matching routes for each query part.
-  auto VaryQueryMatching(const std::vector<SingleQueryPart> &query_parts, const SymbolTable &symbol_table) {
-    std::vector<impl::VaryQueryPartMatching> alternative_query_parts;
-    alternative_query_parts.reserve(query_parts.size());
-    for (const auto &query_part : query_parts) {
-      alternative_query_parts.emplace_back(impl::VaryQueryPartMatching(query_part, symbol_table));
+  auto VaryQueryMatching(const QueryParts &query_parts, const SymbolTable &symbol_table) {
+    std::vector<impl::VaryQueryPartMatching> varying_query_matchings;
+
+    auto single_query_parts = ExtractSingleQueryParts(std::make_unique<QueryParts>(query_parts));
+
+    for (const auto &single_query_part : single_query_parts) {
+      varying_query_matchings.emplace_back(impl::VaryQueryPartMatching(single_query_part, symbol_table));
+    }
+
+    return iter::slice(MakeCartesianProduct(std::move(varying_query_matchings)), 0UL, FLAGS_query_max_plans);
+  }
+
+  std::vector<SingleQueryPart> ExtractSingleQueryParts(const std::shared_ptr<QueryParts> query_parts) {
+    std::vector<SingleQueryPart> results;
+
+    for (const auto &query_part : query_parts->query_parts) {
+      for (const auto &single_query_part : query_part.single_query_parts) {
+        results.push_back(single_query_part);
+
+        for (const auto &subquery : single_query_part.subqueries) {
+          const auto subquery_results = ExtractSingleQueryParts(subquery);
+          results.insert(results.end(), subquery_results.begin(), subquery_results.end());
```

Comment:
```suggestion
          const auto subquery_results = ExtractSingleQueryParts(subquery);
          results.insert(results.end(), std::make_move_iterator(subquery_results.begin()), std::make_move_iterator(subquery_results.end()));
```
Not wrong, but we can avoid copying elements